### PR TITLE
Use std::io::Write to render templates

### PR DESCRIPTION
**Pull request by xabufr**
This allows users to render templates directly to anything that implements std::io::Write (for exemple files).

These changes permits users to control the output buffer when rendering to memory. They should also reduce number of memory allocations, as single rendering steps doesn't return Strings anymore (but instead uses output buffer).

Now the only case where memory really needs to be allocated is when we do not know when the rendered value should really but append to output.

### CODE REVIEW

File: src/renderer/mod.rs [hunk 1]
```diff
@@ -44,15 +47,16 @@ impl<'a> Renderer<'a> {
 
     /// Combines the context with the Template to generate the end result
     pub fn render(&self) -> Result<String> {
-        let output;
-
-        {
-            let mut processor =
-                Processor::new(self.template, self.tera, &self.context, self.should_escape);
+        let mut output = Vec::with_capacity(10_000);
```
Comment (Keats) [inline src/renderer/mod.rs:50]: any reason 10k was chosen?

File: src/renderer/processor.rs [hunk 2]
```diff
@@ -145,21 +147,19 @@ impl<'a> Processor<'a> {
         }
     }
 
-    fn render_body(&mut self, body: &'a [Node]) -> Result<String> {
-        let mut output = String::with_capacity(body.len() * 20);
-
+    fn render_body(&mut self, body: &'a [Node], w: &mut impl Write) -> Result<()> {
```
Comment (Keats) [inline src/renderer/processor.rs:150]: can we rename `w` -> `writer` everwhere?

File: src/utils.rs [hunk 3]
```diff
@@ -33,9 +35,24 @@ pub fn escape_html(input: &str) -> String {
...
+    F: FnOnce(&mut Vec<u8>) -> Result<(), E>,
+    Error: From<E>,
+{
+    let mut buffer = Vec::new();
+    render(&mut buffer).map_err(Error::from)?;
+    Ok(buffer_to_string(buffer))
+}
+
+pub(crate) fn buffer_to_string(buffer: Vec<u8>) -> String {
+    String::from_utf8(buffer).expect("Render only UTF-8 bytes")
```
Comment (Keats) [inline src/utils.rs:49]: let's handle that error as well rather than `expect`

**Review by Keats — CHANGES_REQUESTED** [review]
Looks good.
Can you also add a benchmark for `render_to` to see the difference?

File: src/renderer/mod.rs (hunk 1 above)
Comment (xabufr) [reply src/renderer/mod.rs:50]: The only reason is that previous code allocate a String buffer of 10.000 chars: https://github.com/Keats/tera/pull/601/files#diff-e67fd4dbfd06da05000d6b37dcecbf48e765d6e30eb90d021fa01fae7b094887L1027

File: src/renderer/processor.rs (hunk 2 above)
Comment (xabufr) [reply src/renderer/processor.rs:150]: Yes I will did it

File: src/utils.rs (hunk 3 above)
Comment (xabufr) [reply src/utils.rs:49]: Okay, but this will introduce a new error type (and at the moment this error should never happen as we only write UTF8 bytes)

Comment (Keats) [reply src/utils.rs:49]: It's probably ok, we have non-exhaustive error enums.

File: src/renderer/mod.rs (hunk 1 above)
Comment (Keats) [reply src/renderer/mod.rs:50]: oh I forgot about that

File: src/renderer/processor.rs (hunk 2 above)
Comment (xabufr) [reply src/renderer/processor.rs]: done

File: src/utils.rs (hunk 3 above)
Comment (xabufr) [reply src/utils.rs]: It should be ok now

**Comment by Keats** [comment]
Thanks for the changes! Can you add a benchmark?

**Comment by xabufr** [comment]
Yes I'll try to do that when I had time

**Comment by Keats** [comment]
I'll check to add it myself, thanks!

**Comment by freedeaths** [comment]
Excuse me, when will this "render_to" feature in "next" branch be released in a stable version?

**Comment by Keats** [comment]
Not long, maybe this weekend? Probably a couple more bug fixes to go in https://github.com/Keats/tera/pull/607 if possible

**Pull request merged**